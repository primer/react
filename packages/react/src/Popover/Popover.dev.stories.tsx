import React from 'react'
import type {Meta} from '@storybook/react'
import Heading from '../Heading'
import Popover from './Popover'
import Text from '../Text'
import {Button} from '../Button'

export default {
  title: 'Components/Popover/Dev',
  component: Popover,
} as Meta<typeof Popover>

export const SxProps = () => (
  <Popover
    relative
    open={true}
    caret="top-right"
    sx={{
      left: '50%',
      transform: 'translateX(-50%)',
      mt: 2,
      color: 'var(--bgColor-danger-muted)',
    }}
    style={{padding: '16px'}}
  >
    <Popover.Content
      sx={{
        minWidth: '260px',
        width: '40%',
      }}
      style={{padding: '32px'}}
    >
      <Heading sx={{fontSize: 2}}>Popover heading</Heading>
      <Text as="p">Message about popovers</Text>
      <Button>Got it!</Button>
    </Popover.Content>
  </Popover>
)
