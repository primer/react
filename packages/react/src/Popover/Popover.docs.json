{
  "id": "popover",
  "name": "Popover",
  "status": "alpha",
  "a11yReviewed": false,
  "stories": [],
  "importPath": "@primer/react",
  "props": [
    {
      "name": "as",
      "type": "string",
      "defaultValue": "div",
      "description": "Sets the underlying HTML tag for the component"
    },
    {
      "name": "caret",
      "type": "| 'top' | 'bottom' | 'left' | 'right' | 'bottom-left' | 'bottom-right' | 'top-left' | 'top-right' | 'left-bottom' | 'left-top' | 'right-bottom' | 'right-top' ",
      "defaultValue": "'top'",
      "description": "Controls the position of the caret"
    },
    {
      "name": "open",
      "type": "boolean",
      "defaultValue": "false",
      "description": "Controls the visibility of the popover."
    },
    {
      "name": "relative",
      "type": "boolean",
      "defaultValue": "false",
      "description": "Set to true to render the popover using relative positioning. "
    },
    {
      "name": "sx",
      "type": "SystemStyleObject"
    }
  ],
  "subcomponents": [
    {
      "name": "Popover.Content",
      "props": [
        {
          "name": "as",
          "type": "string",
          "defaultValue": "div",
          "description": "Sets the underlying HTML tag for the component"
        },
        {
          "name": "sx",
          "type": "SystemStyleObject"
        }
      ]
    }
  ]
}