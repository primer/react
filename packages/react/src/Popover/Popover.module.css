.Popover {
  position: absolute;
  z-index: 100;
  display: none;

  &:where([data-open]) {
    display: block;
  }

  &:where([data-relative]) {
    position: relative;
  }
}

.PopoverContent {
  position: relative;
  width: 232px;
  padding: var(--base-size-24);
  margin-right: auto;
  margin-left: auto;
  background-color: var(--overlay-bgColor);
  border: var(--borderWidth-thin) solid var(--borderColor-default);
  border-radius: var(--borderRadius-medium);

  /* Carets */
  &::before,
  &::after {
    position: absolute;
    left: 50%;
    display: inline-block;
    content: '';
  }

  &::before {
    top: calc(-1 * var(--base-size-16));
    /* stylelint-disable-next-line primer/spacing */
    margin-left: -9px;

    /* TODO: solid? */
    /* stylelint-disable-next-line primer/borders */
    border: var(--base-size-8) solid transparent;
    border-bottom-color: var(--borderColor-default);
  }

  &::after {
    /* stylelint-disable-next-line primer/spacing */
    top: -14px;
    margin-left: calc(-1 * var(--base-size-8));

    /* // todo: solid */
    /* stylelint-disable-next-line primer/borders */
    border: 7px solid transparent;
    /* stylelint-disable-next-line primer/colors */
    border-bottom-color: var(--overlay-bgColor);
  }

  /* Bottom-oriented carets */
  :where([data-caret='bottom']) &,
  :where([data-caret='bottom-right']) &,
  :where([data-caret='bottom-left']) & {
    &::before,
    &::after {
      top: auto;
      border-bottom-color: transparent;
    }

    &::before {
      bottom: calc(-1 * var(--base-size-16));
      border-top-color: var(--borderColor-default);
    }

    &::after {
      /* stylelint-disable-next-line primer/spacing */
      bottom: -14px;
      /* stylelint-disable-next-line primer/colors */
      border-top-color: var(--overlay-bgColor);
    }
  }

  /* Top & Bottom: Right-oriented carets */
  :where([data-caret='top-right']) &,
  :where([data-caret='bottom-right']) & {
    /* stylelint-disable-next-line primer/spacing */
    right: -9px;
    margin-right: 0;

    &::before,
    &::after {
      left: auto;
      margin-left: 0;
    }

    &::before {
      /* stylelint-disable-next-line primer/spacing */
      right: 20px;
    }

    &::after {
      /* stylelint-disable-next-line primer/spacing */
      right: 21px;
    }
  }

  /* Top & Bottom: Left-oriented carets */
  :where([data-caret='top-left']) &,
  :where([data-caret='bottom-left']) & {
    /* stylelint-disable-next-line primer/spacing */
    left: -9px;
    margin-left: 0;

    &::before,
    &::after {
      left: var(--base-size-24);
      margin-left: 0;
    }

    &::after {
      /* stylelint-disable-next-line primer/spacing */
      left: calc(var(--base-size-24) + 1px);
    }
  }

  /* Right- & Left-oriented carets */
  :where([data-caret='right']) &,
  :where([data-caret='right-top']) &,
  :where([data-caret='right-bottom']) &,
  :where([data-caret='left']) &,
  :where([data-caret='left-top']) &,
  :where([data-caret='left-bottom']) & {
    &::before,
    &::after {
      top: 50%;
      left: auto;
      margin-left: 0;
      border-bottom-color: transparent;
    }

    &::before {
      /* stylelint-disable-next-line primer/spacing */
      margin-top: calc((var(--base-size-8) + 1px) * -1);
    }

    &::after {
      margin-top: calc(-1 * var(--base-size-8));
    }
  }

  /* Right-oriented carets */
  :where([data-caret='right']) &,
  :where([data-caret='right-top']) &,
  :where([data-caret='right-bottom']) & {
    &::before {
      right: calc(-1 * var(--base-size-16));
      border-left-color: var(--borderColor-default);
    }

    &::after {
      /* stylelint-disable-next-line primer/spacing */
      right: -14px;
      /* stylelint-disable-next-line primer/colors */
      border-left-color: var(--overlay-bgColor);
    }
  }

  /* Left-oriented carets */
  :where([data-caret='left']) &,
  :where([data-caret='left-top']) &,
  :where([data-caret='left-bottom']) & {
    &::before {
      left: calc(-1 * var(--base-size-16));
      border-right-color: var(--borderColor-default);
    }

    &::after {
      /* stylelint-disable-next-line primer/spacing */
      left: -14px;
      /* stylelint-disable-next-line primer/colors */
      border-right-color: var(--overlay-bgColor);
    }
  }

  /* Right & Left: Top-oriented carets */
  :where([data-caret='right-top']) &,
  :where([data-caret='left-top']) & {
    &::before,
    &::after {
      top: var(--base-size-24);
    }
  }

  /* Right & Left: Bottom-oriented carets */
  :where([data-caret='right-bottom']) &,
  :where([data-caret='left-bottom']) & {
    &::before,
    &::after {
      top: auto;
    }

    &::before {
      bottom: var(--base-size-16);
    }

    &::after {
      /* stylelint-disable-next-line primer/spacing */
      bottom: calc(var(--base-size-16) + 1px);
    }
  }
}
