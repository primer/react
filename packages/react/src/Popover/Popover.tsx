import clsx from 'clsx'
import styled from 'styled-components'
import {get} from '../constants'
import type {SxProp} from '../sx'
import sx from '../sx'
import type {ComponentProps} from '../utils/types'

type CaretPosition =
  | 'top'
  | 'bottom'
  | 'left'
  | 'right'
  | 'bottom-left'
  | 'bottom-right'
  | 'top-left'
  | 'top-right'
  | 'left-bottom'
  | 'left-top'
  | 'right-bottom'
  | 'right-top'

type StyledPopoverProps = {
  caret?: CaretPosition
  relative?: boolean
  open?: boolean
} & SxProp

const Popover = styled.div.attrs<StyledPopoverProps>(({className, caret = 'top'}) => {
  return {
    className: clsx(className, `caret-pos--${caret}`),
  }
})<StyledPopoverProps>`
  position: ${props => (props.relative ? 'relative' : 'absolute')};
  z-index: 100;
  display: ${props => (props.open ? 'block' : 'none')};
  ${sx};
`

const PopoverContent = styled.div<SxProp>`
  border: 1px solid ${get('colors.border.default')};
  border-radius: ${get('radii.2')};
  position: relative;
  width: 232px;
  margin-right: auto;
  margin-left: auto;
  padding: ${get('space.4')};
  background-color: ${get('colors.canvas.overlay')};

  // Carets
  &::before,
  &::after {
    position: absolute;
    left: 50%;
    display: inline-block;
    content: '';
  }

  &::before {
    top: -${get('space.3')};
    margin-left: -9px;
    border: ${get('space.2')} solid transparent; // TODO: solid?
    border-bottom-color: ${get('colors.border.default')};
  }

  &::after {
    top: -14px;
    margin-left: -${get('space.2')};
    border: 7px solid transparent; // todo: solid
    border-bottom-color: ${get('colors.canvas.overlay')};
  }

  // Bottom-oriented carets
  ${Popover}.caret-pos--bottom & ,
  ${Popover}.caret-pos--bottom-right & ,
  ${Popover}.caret-pos--bottom-left & {
    &::before,
    &::after {
      top: auto;
      border-bottom-color: transparent;
    }

    &::before {
      bottom: -${get('space.3')};
      border-top-color: ${get('colors.border.default')};
    }

    &::after {
      bottom: -14px;
      // stylelint-disable-next-line primer/borders
      border-top-color: ${get('colors.canvas.overlay')};
    }
  }

  // Top & Bottom: Right-oriented carets
  ${Popover}.caret-pos--top-right & ,
  ${Popover}.caret-pos--bottom-right & {
    right: -9px;
    margin-right: 0;

    &::before,
    &::after {
      left: auto;
      margin-left: 0;
    }

    &::before {
      right: 20px;
    }

    &::after {
      right: 21px;
    }
  }

  // Top & Bottom: Left-oriented carets
  ${Popover}.caret-pos--top-left & ,
  ${Popover}.caret-pos--bottom-left & {
    left: -9px;
    margin-left: 0;

    &::before,
    &::after {
      left: ${get('space.4')};
      margin-left: 0;
    }

    &::after {
      left: calc(${get('space.4')} + 1px);
    }
  }

  // Right- & Left-oriented carets
  ${Popover}.caret-pos--right & ,
  ${Popover}.caret-pos--right-top & ,
  ${Popover}.caret-pos--right-bottom & ,
  ${Popover}.caret-pos--left & ,
  ${Popover}.caret-pos--left-top & ,
  ${Popover}.caret-pos--left-bottom & {
    &::before,
    &::after {
      top: 50%;
      left: auto;
      margin-left: 0;
      border-bottom-color: transparent;
    }

    &::before {
      // stylelint-disable-next-line primer/spacing
      margin-top: calc((${get('space.2')} + 1px) * -1);
    }

    &::after {
      margin-top: -${get('space.2')};
    }
  }

  // Right-oriented carets
  ${Popover}.caret-pos--right & ,
  ${Popover}.caret-pos--right-top & ,
  ${Popover}.caret-pos--right-bottom & {
    &::before {
      right: -${get('space.3')};
      border-left-color: ${get('colors.border.default')};
    }

    &::after {
      right: -14px;
      // stylelint-disable-next-line primer/borders
      border-left-color: ${get('colors.canvas.overlay')};
    }
  }

  // Left-oriented carets
  ${Popover}.caret-pos--left & ,
  ${Popover}.caret-pos--left-top & ,
  ${Popover}.caret-pos--left-bottom & {
    &::before {
      left: -${get('space.3')};
      border-right-color: ${get('colors.border.default')};
    }

    &::after {
      left: -14px;
      // stylelint-disable-next-line primer/borders
      border-right-color: ${get('colors.canvas.overlay')};
    }
  }

  // Right & Left: Top-oriented carets
  ${Popover}.caret-pos--right-top & ,
  ${Popover}.caret-pos--left-top & {
    &::before,
    &::after {
      top: ${get('space.4')};
    }
  }

  // Right & Left: Bottom-oriented carets
  ${Popover}.caret-pos--right-bottom & ,
  ${Popover}.caret-pos--left-bottom & {
    &::before,
    &::after {
      top: auto;
    }

    &::before {
      bottom: ${get('space.3')};
    }

    &::after {
      bottom: calc(${get('space.3')} + 1px);
    }
  }

  ${sx};
`

PopoverContent.displayName = 'Popover.Content'

export type PopoverProps = ComponentProps<typeof Popover>
export type PopoverContentProps = ComponentProps<typeof PopoverContent>
export default Object.assign(Popover, {Content: PopoverContent})
