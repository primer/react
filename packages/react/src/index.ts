'use client'

export {default as theme} from './theme'
export {get as themeGet} from './constants'
export {default as BaseStyles} from './BaseStyles'
export type {BaseStylesProps} from './BaseStyles'
export {default as ThemeProvider, useTheme, useColorSchemeVar} from './ThemeProvider'
export type {ThemeProviderProps} from './ThemeProvider'

// Layout
export {default as Box} from './Box'
export type {BoxProps} from './Box'
export * from './Button'
export {PageLayout} from './PageLayout'
export type {
  PageLayoutProps,
  PageLayoutHeaderProps,
  PageLayoutContentProps,
  PageLayoutPaneProps,
  PageLayoutFooterProps,
} from './PageLayout'
export {SplitPageLayout} from './SplitPageLayout'
export type {
  SplitPageLayoutProps,
  SplitPageLayoutHeaderProps,
  SplitPageLayoutContentProps,
  SplitPageLayoutPaneProps,
  SplitPageLayoutFooterProps,
} from './SplitPageLayout'

// Hooks
export {default as useDetails} from './hooks/useDetails'
export {default as useSafeTimeout} from './hooks/useSafeTimeout'
export {useOnOutsideClick} from './hooks/useOnOutsideClick'
export type {TouchOrMouseEvent} from './hooks/useOnOutsideClick'
export {useOpenAndCloseFocus} from './hooks/useOpenAndCloseFocus'
export {useOnEscapePress} from './hooks/useOnEscapePress'
export {useOverlay} from './hooks/useOverlay'
export {useConfirm} from './ConfirmationDialog/ConfirmationDialog'
export {useFocusTrap} from './hooks/useFocusTrap'
export type {FocusTrapHookSettings} from './hooks/useFocusTrap'
export {useFocusZone} from './hooks/useFocusZone'
export type {FocusZoneHookSettings} from './hooks/useFocusZone'
export {useRefObjectAsForwardedRef} from './hooks/useRefObjectAsForwardedRef'
export {useResizeObserver} from './hooks/useResizeObserver'
export {useResponsiveValue} from './hooks/useResponsiveValue'

// Utils
export {createComponent} from './utils/create-component'

// Components
export {default as Radio} from './Radio'
export type {RadioProps} from './Radio'
export {ActionList} from './ActionList'
export type {
  ActionListProps,
  ActionListGroupProps,
  ActionListItemProps,
  ActionListLinkItemProps,
  ActionListDividerProps,
  ActionListDescriptionProps,
  ActionListLeadingVisualProps,
  ActionListTrailingVisualProps,
} from './ActionList'
export {ActionMenu} from './ActionMenu'
export type {ActionMenuProps, ActionMenuAnchorProps, ActionMenuButtonProps} from './ActionMenu'
export {AnchoredOverlay} from './AnchoredOverlay'
export type {AnchoredOverlayProps} from './AnchoredOverlay'
export {default as Autocomplete} from './Autocomplete'
export type {AutocompleteMenuProps, AutocompleteInputProps, AutocompleteOverlayProps} from './Autocomplete'
export {default as Avatar} from './Avatar'
export type {AvatarProps} from './Avatar'
export {default as AvatarPair} from './AvatarPair'
export type {AvatarPairProps} from './AvatarPair'
export {default as AvatarStack} from './AvatarStack'
export type {AvatarStackProps} from './AvatarStack'
export {default as BranchName} from './BranchName'
export type {BranchNameProps} from './BranchName'
export {default as Breadcrumbs, Breadcrumb} from './Breadcrumbs'
export type {BreadcrumbsProps, BreadcrumbsItemProps, BreadcrumbProps, BreadcrumbItemProps} from './Breadcrumbs'
export {default as ButtonGroup} from './ButtonGroup'
export type {ButtonGroupProps} from './ButtonGroup'
export {default as Caret} from './Caret'
export type {CircleBadgeProps, CircleBadgeIconProps} from './CircleBadge'
export {default as CircleOcticon} from './CircleOcticon'
export type {CircleOcticonProps} from './CircleOcticon'
export {default as CheckboxGroup} from './CheckboxGroup'
export type {CaretProps} from './Caret'
export {default as CircleBadge} from './CircleBadge'
export {default as CounterLabel} from './CounterLabel'
export type {CounterLabelProps} from './CounterLabel'
export {default as Details} from './Details'
export type {DetailsProps} from './Details'
export {default as Dialog} from './Dialog'
export type {DialogProps, DialogHeaderProps} from './Dialog'
export type {ConfirmationDialogProps} from './ConfirmationDialog/ConfirmationDialog'
export {ConfirmationDialog} from './ConfirmationDialog/ConfirmationDialog'
export {default as Flash} from './Flash'
export type {FlashProps} from './Flash'
export {default as FormControl} from './FormControl'
export {useFormControlForwardedProps} from './FormControl'
export {default as Header} from './Header'
export type {HeaderProps, HeaderItemProps, HeaderLinkProps} from './Header'
export {default as Heading} from './Heading'
export type {HeadingProps} from './Heading'
export {default as Label} from './Label'
export type {LabelProps} from './Label'
export {default as LabelGroup} from './LabelGroup'
export type {LabelGroupProps} from './LabelGroup'
export {default as Link} from './Link'
export type {LinkProps} from './Link'
export {NavList} from './NavList'
export type {
  NavListProps,
  NavListItemProps,
  NavListSubNavProps,
  NavListGroupProps,
  NavListLeadingVisualProps,
  NavListTrailingVisualProps,
  NavListDividerProps,
} from './NavList'
export {default as Octicon} from './Octicon'
export type {OcticonProps} from './Octicon'
export {default as Overlay} from './Overlay'
export type {OverlayProps} from './Overlay'
export {default as Pagehead} from './Pagehead'
export type {PageheadProps} from './Pagehead'
export {default as Pagination} from './Pagination'
export type {PaginationProps} from './Pagination'
export {default as PointerBox} from './PointerBox'
export type {PointerBoxProps} from './PointerBox'
export {default as Popover} from './Popover'
export type {PopoverProps, PopoverContentProps} from './Popover'
export {default as Portal, registerPortalRoot} from './Portal'
export type {PortalProps} from './Portal'
export {ProgressBar} from './ProgressBar'
export type {ProgressBarProps} from './ProgressBar'
export {default as RadioGroup} from './RadioGroup'
export type {RelativeTimeProps} from './RelativeTime'
export {default as RelativeTime} from './RelativeTime'
export {SegmentedControl} from './SegmentedControl'
export {default as Select} from './Select'
export type {SelectProps} from './Select'
export {SelectPanel} from './SelectPanel'
export type {SelectPanelProps} from './SelectPanel'
export {default as SideNav} from './SideNav'
export type {SideNavProps, SideNavLinkProps} from './SideNav'
export {default as Spinner} from './Spinner'
export type {SpinnerProps} from './Spinner'
export {default as StateLabel} from './StateLabel'
export type {StateLabelProps} from './StateLabel'
export {default as SubNav} from './SubNav'
export type {SubNavProps, SubNavLinkProps, SubNavLinksProps} from './SubNav'
export {default as ToggleSwitch} from './ToggleSwitch'
export {default as TabNav} from './TabNav'
export type {TabNavProps, TabNavLinkProps} from './TabNav'
export {default as TextInput} from './TextInput'
export type {TextInputProps} from './TextInput'
export {default as TextInputWithTokens} from './TextInputWithTokens'
export type {TextInputWithTokensProps} from './TextInputWithTokens'
export {default as Text} from './Text'
export type {TextProps} from './Text'
export {default as Timeline} from './Timeline'
export type {
  TimelineProps,
  TimelineBadgeProps,
  TimelineBodyProps,
  TimelineBreakProps,
  TimelineItemsProps,
} from './Timeline'
export {default as Token, IssueLabelToken, AvatarToken} from './Token'
export type {TokenProps} from './Token'
export {default as Tooltip} from './Tooltip/Tooltip'
export type {TooltipProps} from './Tooltip/Tooltip'
export {default as Truncate} from './Truncate'
export type {TruncateProps} from './Truncate'

export {default as Checkbox} from './Checkbox'
export type {CheckboxProps} from './Checkbox'

export {default as Textarea} from './Textarea'
export type {TextareaProps} from './Textarea'

export {TreeView} from './TreeView'
export type {
  TreeViewProps,
  TreeViewItemProps,
  TreeViewSubTreeProps,
  TreeViewVisualProps,
  TreeViewErrorDialogProps,
} from './TreeView'

export {UnderlineNav} from './UnderlineNav'
export type {UnderlineNavProps, UnderlineNavItemProps} from './UnderlineNav'

export {ActionBar} from './ActionBar'
export type {ActionBarProps} from './ActionBar'

// eslint-disable-next-line no-restricted-imports
export {SSRProvider, useSSRSafeId} from './utils/ssr'
export {default as sx, merge} from './sx'
export type {SxProp} from './sx'
