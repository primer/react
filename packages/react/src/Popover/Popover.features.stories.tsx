import type {Meta, StoryFn} from '@storybook/react-vite'
import Heading from '../Heading'
import Popover from './Popover'
import Text from '../Text'
import {Button} from '../Button'
import React from 'react'

export default {
  title: 'Components/Popover/Features',
  component: Popover,
} as Meta<typeof Popover>

export const CloseOnClickOutside = () => {
  const [open, setOpen] = React.useState(true)
  const buttonRef = React.useRef<HTMLButtonElement>(null)
  return (
    <>
      <Button
        style={{marginLeft: 'auto', marginRight: 'auto', marginBottom: 'var(--base-size-4)'}}
        ref={buttonRef}
        onClick={() => setOpen(prev => !prev)}
      >
        Toggle Popover
      </Button>
      <Popover relative open={open} caret="top">
        <Popover.Content
          style={{marginTop: 'var(--base-size-8)'}}
          onClickOutside={() => setOpen(false)}
          ignoreClickRefs={[buttonRef]}
        >
          <Heading style={{fontSize: 'var(--text-title-size-small)'}}>Popover heading</Heading>
          <Text as="p">Message about popovers</Text>
          <Button>Got it!</Button>
        </Popover.Content>
      </Popover>
    </>
  )
}
