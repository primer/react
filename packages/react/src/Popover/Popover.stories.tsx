import React from 'react'
import type {Meta, StoryFn} from '@storybook/react'
import Heading from '../Heading'
import Popover from './Popover'
import Text from '../Text'
import {Button} from '../Button'

export default {
  title: 'Components/Popover',
  component: Popover,
} as Meta<typeof Popover>

export const Default = () => (
  <Popover relative open={true} caret="top">
    <Popover.Content sx={{marginTop: 2}}>
      <Heading sx={{fontSize: 2}}>Popover heading</Heading>
      <Text as="p">Message about popovers</Text>
      <Button>Got it!</Button>
    </Popover.Content>
  </Popover>
)

export const Playground: StoryFn<typeof Popover> = args => (
  <Popover {...args}>
    <Popover.Content sx={{marginTop: 2}}>
      <Heading sx={{fontSize: 2}}>Popover heading</Heading>
      <Text as="p">Message about popovers</Text>
      <Button>Got it!</Button>
    </Popover.Content>
  </Popover>
)

Playground.args = {
  caret: 'top',
  open: true,
  relative: true,
}

Playground.argTypes = {
  caret: {
    control: {
      type: 'radio',
    },
    options: [
      'top',
      'bottom',
      'left',
      'right',
      'bottom-left',
      'bottom-right',
      'top-left',
      'top-right',
      'left-bottom',
      'left-top',
      'right-bottom',
      'right-top',
    ],
  },
  open: {
    control: {
      type: 'boolean',
    },
  },
  relative: {
    control: {
      type: 'boolean',
    },
  },
}
