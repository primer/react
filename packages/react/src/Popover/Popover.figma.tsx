import React from 'react'
import {Heading, Popover, Text} from '../../src'
import figma from '@figma/code-connect'

/**
 * -- This file was auto-generated by `figma connect create` --
 * `props` includes a mapping from Figma properties and variants to
 * suggested values. You should update this to match the props of your
 * code component, and update the `example` function to return the
 * code example you'd like to see in Figma
 */

figma.connect(Popover, 'https://www.figma.com/design/GCvY3Qv8czRgZgvl1dG6lp/Primer-Web?node-id=19015-63778&m=dev', {
  props: {
    body: figma.string('body'),
    heading: figma.string('heading'),
    action: figma.children('Action'),
    caret: figma.enum('caret', {
      top: 'top',
      'top-left': 'top-left',
      'top-right': 'top-right',
      bottom: 'bottom',
      'bottom-left': 'bottom-left',
      'bottom-right': 'bottom-right',
      left: 'left',
      'left-top': 'left-top',
      'left-bottom': 'left-bottom',
      right: 'right',
      'right-top': 'right-top',
      'right-bottom': 'right-bottom',
    }),
  },
  example: ({caret, heading, body, action}) => (
    <Popover caret={caret}>
      <Popover.Content>
        <Heading
          sx={{
            fontSize: 2,
          }}
        >
          {heading}
        </Heading>
        <Text as="p">{body}</Text>
        {action}
      </Popover.Content>
    </Popover>
  ),
})
