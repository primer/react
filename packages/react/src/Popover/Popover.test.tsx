import {describe, expect, it} from 'vitest'
import {render} from '@testing-library/react'
import type {PopoverProps} from '../Popover'
import Popover from '../Popover'

describe('Popover', () => {
  it('should support `className` on the outermost element', () => {
    const Element = () => <Popover className={'test-class-name'}></Popover>
    expect(render(<Element />).container.firstChild).toHaveClass('test-class-name')
  })

  const CARET_POSITIONS: PopoverProps['caret'][] = [
    'top',
    'bottom',
    'left',
    'right',
    'bottom-left',
    'bottom-right',
    'top-left',
    'top-right',
    'left-bottom',
    'left-top',
    'right-bottom',
    'right-top',
  ]

  for (const pos of CARET_POSITIONS) {
    it(`renders correctly for a caret position of ${pos}`, () => {
      const element = (
        <Popover caret={pos} open>
          <Popover.Content>Hello!</Popover.Content>
        </Popover>
      )

      const {container} = render(element)
      expect(container.firstChild).toHaveAttribute('data-caret', pos)
    })
  }

  it('renders both elements as a <div>', () => {
    const {container: popoverContainer} = render(<Popover />)
    const {container: contentContainer} = render(<Popover.Content />)
    expect((popoverContainer.firstChild as Element).tagName).toEqual('DIV')
    expect((contentContainer.firstChild as Element).tagName).toEqual('DIV')
  })
})
