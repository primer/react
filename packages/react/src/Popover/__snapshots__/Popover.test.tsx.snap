// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover renders correctly for a caret position of bottom 1`] = `
<div
  className="Popover"
  data-caret="bottom"
  data-open=""
>
  <div
    className="PopoverContent"
    data-height="fit-content"
    data-overflow="auto"
    data-width="small"
  >
    Hello!
  </div>
</div>
`;

exports[`Popover renders correctly for a caret position of bottom-left 1`] = `
<div
  className="Popover"
  data-caret="bottom-left"
  data-open=""
>
  <div
    className="PopoverContent"
    data-height="fit-content"
    data-overflow="auto"
    data-width="small"
  >
    Hello!
  </div>
</div>
`;

exports[`Popover renders correctly for a caret position of bottom-right 1`] = `
<div
  className="Popover"
  data-caret="bottom-right"
  data-open=""
>
  <div
    className="PopoverContent"
    data-height="fit-content"
    data-overflow="auto"
    data-width="small"
  >
    Hello!
  </div>
</div>
`;

exports[`Popover renders correctly for a caret position of left 1`] = `
<div
  className="Popover"
  data-caret="left"
  data-open=""
>
  <div
    className="PopoverContent"
    data-height="fit-content"
    data-overflow="auto"
    data-width="small"
  >
    Hello!
  </div>
</div>
`;

exports[`Popover renders correctly for a caret position of left-bottom 1`] = `
<div
  className="Popover"
  data-caret="left-bottom"
  data-open=""
>
  <div
    className="PopoverContent"
    data-height="fit-content"
    data-overflow="auto"
    data-width="small"
  >
    Hello!
  </div>
</div>
`;

exports[`Popover renders correctly for a caret position of left-top 1`] = `
<div
  className="Popover"
  data-caret="left-top"
  data-open=""
>
  <div
    className="PopoverContent"
    data-height="fit-content"
    data-overflow="auto"
    data-width="small"
  >
    Hello!
  </div>
</div>
`;

exports[`Popover renders correctly for a caret position of right 1`] = `
<div
  className="Popover"
  data-caret="right"
  data-open=""
>
  <div
    className="PopoverContent"
    data-height="fit-content"
    data-overflow="auto"
    data-width="small"
  >
    Hello!
  </div>
</div>
`;

exports[`Popover renders correctly for a caret position of right-bottom 1`] = `
<div
  className="Popover"
  data-caret="right-bottom"
  data-open=""
>
  <div
    className="PopoverContent"
    data-height="fit-content"
    data-overflow="auto"
    data-width="small"
  >
    Hello!
  </div>
</div>
`;

exports[`Popover renders correctly for a caret position of right-top 1`] = `
<div
  className="Popover"
  data-caret="right-top"
  data-open=""
>
  <div
    className="PopoverContent"
    data-height="fit-content"
    data-overflow="auto"
    data-width="small"
  >
    Hello!
  </div>
</div>
`;

exports[`Popover renders correctly for a caret position of top 1`] = `
<div
  className="Popover"
  data-caret="top"
  data-open=""
>
  <div
    className="PopoverContent"
    data-height="fit-content"
    data-overflow="auto"
    data-width="small"
  >
    Hello!
  </div>
</div>
`;

exports[`Popover renders correctly for a caret position of top-left 1`] = `
<div
  className="Popover"
  data-caret="top-left"
  data-open=""
>
  <div
    className="PopoverContent"
    data-height="fit-content"
    data-overflow="auto"
    data-width="small"
  >
    Hello!
  </div>
</div>
`;

exports[`Popover renders correctly for a caret position of top-right 1`] = `
<div
  className="Popover"
  data-caret="top-right"
  data-open=""
>
  <div
    className="PopoverContent"
    data-height="fit-content"
    data-overflow="auto"
    data-width="small"
  >
    Hello!
  </div>
</div>
`;
